cocos/editor-support/cocostudio/WidgetReader/GameNode3DReader/GameNode3DReader.cpp

```diff
@@ -0,0 +1,427 @@
+                                builder->CreateString(leftPath),
+                                builder->CreateString(leftPlistFile),
+                                leftResourceType),
+            CreateResourceData(*builder,
+                                builder->CreateString(rightPath),
+                                builder->CreateString(rightPlistFile),
+                                rightResourceType),
+            CreateResourceData(*builder,
+                                builder->CreateString(upPath),
+                                builder->CreateString(upPlistFile),
+                                upResourceType),
+            CreateResourceData(*builder,
+                                builder->CreateString(downPath),
+                                builder->CreateString(downPlistFile),
+                                downResourceType),
+            CreateResourceData(*builder,
+                                builder->CreateString(forwardPath),
+                                builder->CreateString(forwardPlistFile),
+                                forwardResourceType),
+            CreateResourceData(*builder,
+                                builder->CreateString(backPath),
+                                builder->CreateString(backPlistFile),
+                                backResourceType),
+            builder->CreateString(frameEvent),
+            builder->CreateString(customProperty)
+            );
+
+        return *(Offset<Table>*)(&options);
+    }
+    
+    void GameNode3DReader::setPropsWithFlatBuffers(cocos2d::Node *node,
+                                                   const flatbuffers::Table* node3DOptions)
+    {
+        auto options = (GameNode3DOption*)node3DOptions;
+        
+        std::string name = options->name()->c_str();
+        node->setName(name);
+
+        bool skyBoxEnabled = options->skyBoxEnabled() != 0;
+        if (skyBoxEnabled)
+        {
+            auto leftFileData = options->leftFileData()->path()->c_str();
+            auto rightFileData = options->rightFileData()->path()->c_str();
+            auto upFileData = options->upFileData()->path()->c_str();
+            auto downFileData = options->downFileData()->path()->c_str();
+            auto forwardFileData = options->forwardFileData()->path()->c_str();
+            auto backFileData = options->backFileData()->path()->c_str();
+
+            Skybox* childBox = Skybox::create(leftFileData,rightFileData,upFileData,downFileData,forwardFileData,backFileData);
+            unsigned short cameraFlag = 1 << 10;
+            childBox->setCameraMask(cameraFlag, false);
+            node->addChild(childBox,0,"_innerSkyBox");
+        }
+
+        std::string customProperty = options->customProperty()->c_str();
+        ObjectExtensionData* extensionData = ObjectExtensionData::create();
+        extensionData->setCustomProperty(customProperty);
+        node->setUserObject(extensionData);
```

Comment:
The parser logic is internal and transparent to developers, so I don't think we can use UserObject to store any data for internal usage. If we do so, setUserObject as a public function is dangerous, developer may overwrite the data without knowing it, and it will cause bugs which are very hard to track.

I suggest to use Component to store such data